bundle/Service/ScheduledVisibilityService.php

```diff
@@ -0,0 +1,155 @@
+use function in_array;
+use function sprintf;
+
+final class ScheduledVisibilityService
+{
+    public function __construct(
+        private readonly iterable $handlers,
+        private readonly string $type,
+        private readonly bool $enabled,
+        private readonly bool $allContentTypes,
+        private readonly array $allowedContentTypes,
+    ) {}
+
+    /**
+     * @throws OutOfBoundsException
+     */
+    public function toggleVisibility(Content $content): VisibilityAction
+    {
+        if (!$this->accept($content)) {
+            return VisibilityAction::NoChange;
+        }
+
+        $handlerType = HandlerType::from($this->type);
+
+        /** @var ScheduledVisibilityInterface $handler */
+        foreach ($this->handlers as $handler) {
+            if ($handler->getType() === $handlerType) {
+                if ($this->shouldHide($content)) {
+                    $handler->hide($content);
+
+                    return VisibilityAction::Hidden;
+                }
+                if ($this->shouldReveal($content)) {
+                    $handler->reveal($content);
+
+                    return VisibilityAction::Revealed;
+                }
+
+                return VisibilityAction::NoChange;
+            }
+        }
+
+        throw new OutOfBoundsException(
+            sprintf('Scheduled visibility handler %s could not be resolved', $this->type),
+        );
+    }
+
+    public function accept(Content $content): bool
+    {
+        $enabled = $this->enabled;
+        if (!$enabled) {
+            return false;
+        }
+
+        $allowedAll = $this->allContentTypes;
+        $allowedContentTypes = $this->allowedContentTypes;
+        $contentType = $content->getContentType();
+        if (!$allowedAll && !in_array($contentType->identifier, $allowedContentTypes, true)) {
+            return false;
+        }
+
+        $fieldDefinitions = $contentType->getFieldDefinitions();
+        if (!$fieldDefinitions->has('publish_from') || !$fieldDefinitions->has('publish_to')) {
+            return false;
+        }
+
+        /** @var Field $publishFromField */
+        $publishFromField = $content->getField('publish_from');
+
+        /** @var Field $publishToField */
+        $publishToField = $content->getField('publish_to');
+        if (
+            ($publishFromField->fieldTypeIdentifier !== 'ezdatetime' && $publishFromField->fieldTypeIdentifier !== 'ezdate')
+            || ($publishToField->fieldTypeIdentifier !== 'ezdatetime' && $publishToField->fieldTypeIdentifier !== 'ezdate')
+        ) {
+            return false;
+        }
+
+        $publishFromDateTime = $this->getDateTime($publishFromField);
+        $publishToDateTime = $this->getDateTime($publishToField);
+        if ($publishFromDateTime !== null && $publishToDateTime !== null && $publishToDateTime <= $publishFromDateTime) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public function shouldHide(Content $content): bool
```

Comment:
Since this method doesn't really decide if the hiding should be applied on the `$content`, rename to `shouldBeHidden()`. The we can write:

```php
if ($this->shouldBeHidden($content) && !$handler->isHidden($content)) {
    $handler->hide($content);
}
```

And:

```php
if ($this->shouldBeVisible($content) && !$handler->isVisible($content)) {
    $handler->reveal($content);
}
```

Reply:
fixed in ca56a4a66291d86a4aa11f521c66b2cc28901480